### PR TITLE
[Workspace] fix apps are missing when updating a workspace

### PR DESCRIPTION
**Pull request by ruanyl**
This is caused by #6234 which marked apps as inaccessible when the apps are not configured into the current workspace. However, inaccessible apps are filter out on workspace update page, this is unexpected as such apps can be configured into the current workspace again though they are not configured into the workspace at the moment.

### Issues Resolved
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6481

## Changelog
- skip

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by SuZhou-Joe — APPROVED** [review]
No much concern but maybe we should consider using capabilities service to filter out apps in workspace in the future.

File: src/plugins/workspace/public/components/workspace_form/utils.ts [hunk 1]
```diff
@@ -44,18 +44,28 @@ export const getNumberOfErrors = (formErrors: WorkspaceFormErrors) => {
...
 
   // Filter out all hidden applications and management applications and default selected features
-  const visibleApplications = applications.filter(
-    ({ navLinkStatus, chromeless, category, id }) =>
+  const visibleApplications = applications.filter(({ navLinkStatus, chromeless, category, id }) => {
+    /**
+     * Restrict apps are apps that can be configured into a workspace, but restrict to access
+     * because the current workspace didn't have the apps configured, such apps should NOT filter out
+     */
+    if (restrictedApps && restrictedApps.has(id)) {
```
Comment (wanglam) [inline src/plugins/workspace/public/components/workspace_form/utils.ts:58]: In order to keep the same source in workspace create, update and list page, can we use `restrictedApps` for all customized features for workspaces. We can remove the check below.
```ts
  return (
        navLinkStatus !== AppNavLinkStatus.hidden &&
        !chromeless &&
        !DEFAULT_SELECTED_FEATURES_IDS.includes(id) &&
        category?.id !== DEFAULT_APP_CATEGORIES.management.id
    );
```

**Review by SuZhou-Joe — DISMISSED** [review]
LGTM

**Comment by ZilongX** [comment]
picked up latest main and re-running all checks now

**Review by ZilongX — APPROVED** [review]

File: src/plugins/workspace/public/components/workspace_form/utils.ts [hunk 2]
```diff
@@ -65,23 +61,14 @@ export const convertApplicationsToFeaturesOrGroups = (
 ) => {
```
Comment (wanglam) [inline src/plugins/workspace/public/components/workspace_form/utils.ts:61]: It seems the applications only need id, title and category. Can we remove navLinkStatus and chromeless?

File: src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx [hunk 3]
```diff
@@ -19,21 +19,20 @@ import { PublicAppInfo } from '../../../../../core/public';
 import { isWorkspaceFeatureGroup, convertApplicationsToFeaturesOrGroups } from './utils';
 
 export interface WorkspaceFeatureSelectorProps {
-  applications: Array<
-    Pick<PublicAppInfo, 'id' | 'title' | 'category' | 'chromeless' | 'navLinkStatus'>
-  >;
   selectedFeatures: string[];
   onChange: (newFeatures: string[]) => void;
+  workspaceConfigurableApps?: PublicAppInfo[];
```
Comment (wanglam) [inline src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx:24]: Seems we only need id, title and category for workspaceConfigurableApps here. Could we change to `Pick<PublicAppInfo, 'id' | 'title' | 'category'>`?

File: src/plugins/workspace/public/utils.ts [hunk 4]
```diff
@@ -99,3 +107,16 @@ export function isAppAccessibleInWorkspace(app: App, workspace: WorkspaceObject)
   }
   return false;
 }
+
+export const filterWorkspaceConfigurableApps = (applications: PublicAppInfo[]) => {
+  const visibleApplications = applications.filter(({ navLinkStatus, chromeless, category, id }) => {
+    return (
```
Comment (wanglam) [inline src/plugins/workspace/public/utils.ts:113]: Shall we add some unit tests here to cover this line?

Comment (ruanyl) [reply src/plugins/workspace/public/utils.ts:113]: Yes, there will be a followup PR from @yubonluo to cover this.

File: src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx (hunk 3 above)
Comment (ruanyl) [reply src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx:24]: Actually I don't see a strong need to turn it into `Pick<PublicAppInfo, 'id' | 'title' | 'category'>` in this case. I understood that passing minimum set of data to child components is a best practice, but the `application ` which type is `PublicAppInfo` is provided by the core module, the components expect `PublicAppInfo` type of data sounds reasonable to me as well.

I'm not holding a strong opinion, what else concerns do you have?

**Review by SuZhou-Joe — APPROVED** [review]

File: src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx (hunk 3 above)
Comment (wanglam) [reply src/plugins/workspace/public/components/workspace_form/workspace_feature_selector.tsx:24]: Not a big issue. We can keep current implementation.

**Comment by ZilongX** [comment]
pick up latest main and re-running all checks now

**Review by ZilongX — APPROVED** [review]

**Pull request merged**